crush: add devices class that rules can use as a filter

dachary opened the pull request:
http://tracker.ceph.com/issues/18943

liewegas commented on src/crush/grammar.h:
```diff
@@ -108,7 +108,7 @@ struct crush_grammar : public grammar<crush_grammar>
 
       // devices
-      device = str_p("device") >> posint >> name;
+      device = str_p("device") >> posint >> name >> *( str_p("class") >> name );
```
i think * should be ! (* = 0 or more, ! = 0 or 1)

dachary replied:
done

liewegas commented on src/crush/CrushCompiler.cc:
```diff
@@ -361,7 +361,15 @@ int CrushCompiler::parse_device(iter_t const& i)
+    if (verbose) err << " class " << " '" << c << "'" << std::endl;
+  } else {
+    if (verbose) err << std::endl;
+  }
```
maybe add the decompile case to this patch too, so that a decompile -> compile cycle will not lose informaiton?

dachary replied:
The compilation phase still needs to duplicate the crush trees that the de-compilation will need to guess device classes are in use. It is my understanding that the encoded crush map does not know anything about device classes. It is inferred from the naming of the buckets that contains a special character. Am I going in the wrong direction ?

liewegas commented on src/crush/helpers.h:
```diff
@@ -0,0 +1,27 @@
+ * 
+ * @returns the size of __buckets__ on success, < 0 on error
+ */
+extern int crush_find_roots(struct crush_map *map, int **buckets);
```
void find_roots(set<int>& roots) const;
already exists in CrushWrapper

liewegas replied:
if the intention is to have a C implementation, we should consolidate so that that one calls this one.

dachary replied:
I had no intention, I just did not look... thanks :-)

liewegas commented:
we should include the device classes in the encoded crush map.  clients 
won't need to understand/decode it, but it should be stored as part of the 
canonical map.

for just just stick it at the end of the encoding.  we should maybe 
reserve some space here for future tunables just so clients don't have to 
parse this device class info, but that's just an optimization.

dachary commented:
I'm unclear how we can add to the existing crushmap encoding. Is there versioning that I don't see ? Or is it via feature bits ?

liewegas commented:
anything at the end is ignored by old code, so just append anything new 
there.

dachary commented:
and what happens when the new code reads an old crushmap ?

liewegas commented:
i decodes as much as it expects/understands, then ignores the rest.

liewegas commented:
oh sorry, that's old new.  code code will check if it's reached the end 
of the buffer for any optional parts of the map.  see all the if 
(!p.end()) checsk in decode()

dachary commented:
got it, the code speaks for itself indeed. I got confused but now it's clear :-)

dachary commented:
Should we create a tree for each root and each device class ? Or is it enough to create a tree for each bucket that is referenced by a TAKE rule step ?

dachary commented:
updated  **crush: parse "class XXX" after device** to compile/decompile as suggested

dachary commented:
@liewegas de2fbe1 is a rough draft implementation that does not even compile. It would be great if you could take a quick look and see if something is obviously going in the wrong direction. Here is what it does:

- on finalize() create a new bucket tree for each root and each device class, with buckets named like **name~deviceclass**
- the bucket tree cloning function mirrors the hierarchy and only keeps devices of the specified class
- after the bucket tree have been cloned per device class, the rules that have a take operator referencing a device class are modified to reference the new bucket tree (this is a little trickier than it should be since the ruleno is exposed to the user and the modified rule must take the place of the original rule)
- encode() calls finalize() to ensure the generated buckets reflect the last modifications (this is for OSDMonitor that copies crush maps with encode/decode, other callers know to call finalize after modifying the crush map)

To ensure the crush map consistency, finalize() removes all generated rules / buckets before creating them.

liewegas commented on src/crush/CrushCompiler.cc:
```diff
@@ -658,7 +663,25 @@ int CrushCompiler::parse_rule(iter_t const& i)
+            err << " class " << class_name << std::endl;
+        }
+
+	crush.set_rule_step_take(ruleno, step++, item_id[item], c);
```
I think ehre we actaully want to map item to the parallel per-class tree.

We could also pass c into the rule step since it's ignored, but i'd avoid doing that until we have a reason to use it within the rule processing itself..

liewegas commented on src/crush/CrushCompiler.cc:
```diff
@@ -240,14 +244,28 @@ int CrushCompiler::decompile(ostream &out)
+    // Use ri instead of i here: we need the ruleno of the generated
+    // rule because it is exposed to the user. The original rule is
+    // given another ruleno, never exposed to the user.
+    out << "\truleset " << crush.get_rule_mask_ruleset(ri) << "\n";
```
do we actually need to generate new rules?  i think the original rule can simply use the generated item for the take call, and when decompiled, the fact that the item is generated will tell us to add the 'class xxx' argument to the take step.

dachary replied:
good idea, I'll try that, thanks !

dachary commented:
@liewegas I'm happy about this series but it's still fresh and I may have overlooked a few things. Hopefully not something big but I'm not yet 100% sure. 

I think the goal for this pull request should be to implement and document the "class XXX" for devices & step take only. At this point the users would have no control and would not even be able to see the modified rule step or the generated buckets. But that new feature will save them from the pain of maintaining two parallel trees. Displaying meaningful information about the generated trees via ceph osd crush tree etc. could be implemented afterwards.

liewegas commented on src/crush/CrushWrapper.cc:
```diff
@@ -1167,6 +1167,66 @@ int CrushWrapper::remove_rule(int ruleno)
+					 original->hash,
+					 original->type,
+					 0, NULL, NULL);
+  if(copy == NULL)
```
whitespace

liewegas commented:
Somewhere we also need a per-bucket map of class -> id so that, given a bucket x, we can find the derivative for a given class.  That's needed so that when we decompile we can supplement the 'id NNN' line with 'class-id <class> <id>' (or similar) so that the derivative bucket ids are preserved across decompile->compile.

If that's in place, I think some of the other parts would be simpler:

* after compiling the buckets, we can generate per-class buckets for all roots and all classes, fully populating the per-bucket class->id map.
* the rule compile can then just do that translation as it goes, without the map of ruleno -> step -> class

?

dachary commented:
> the derivative bucket ids are preserved across decompile->compile

Since the derivative bucket id is never exposed, why would we need to preserve it ?

liewegas commented:
because it's an input to teh crash choose hash function.. if it changes 
between crush map versions the pgs will be randomly redistributed for that 
bucket

dachary commented:
@liewegas I overlooked that, thanks for explaining.

dachary commented:
@liewegas this is hopefully fixed and it indeed makes things simpler in some places (see c2ab0c9 for the draft implementation, passes manual tests). The syntax is a little different from what you proposed originally to be consistent with the rest:
```
host host1 {
	id -2		# do not change unnecessarily
	id -7 class ssd		# do not change unnecessarily
	id -12 class hdd		# do not change unnecessarily
	# weight 1.000
	alg straw
	hash 0	# rjenkins1
	item device1 weight 1.000
}
```
It would be great if you could review this draft before I clean it up, in case something significant is missing :-)

liewegas commented on src/crush/CrushCompiler.cc:
```diff
@@ -673,23 +721,35 @@ int CrushCompiler::parse_rule(iter_t const& i)
+          if (crush.class_bucket.count(id) == 0) {
+            err << "in rule '" << rname << "' step take " << item
+                << " has no class information" << std::endl;
+            return -EINVAL;
```
this is a bug, right?  we should have ids for all classes and all buckets at this stage?

dachary replied:
yes, it could be an assert

dachary commented on src/crush/CrushCompiler.cc:
```diff
@@ -603,6 +647,9 @@ int CrushCompiler::parse_bucket(iter_t const& i)
 
+  for (auto &i : class_id)
+    crush.class_bucket[id][i.first] = i.second;
+
```
cool

dachary replied:
we don't want to replace all of crush.class_bucket[id], just add to it.

dachary commented:
@liewegas I thought about how things will be with this implementation from the user perspective (human & not human). Since it's not transparent, ceph osd crush tree will show everything, including the generated buckets & modified rules. That does not strike me as a problem. However the user may be in a position to do a number of things that lead to inconsistency using the **ceph osd crush...** commands.

For the osd & ceph clients who are not aware of the device classes, all should be ok since the crush map is encoded with all the generated buckets & substituted rule steps. Decoding does not expect anything new.

Does that seem right ?

dachary commented:
jenkins test this please (log expired)

dachary commented:
<code>teuthology-suite -k distro --priority 1000 --suite rados --subset $(expr $RANDOM % 50)/50 --email loic@dachary.org --ceph wip-18943-device-class --machine-type smithi</code>

* **running** http://pulpito.ceph.com/loic-2017-02-27_14:30:21-rados-wip-18943-device-class-distro-basic-smithi/

dachary commented:
@liewegas I think it's ready for a detailed review. I'm running a rados suite to catch border cases. I don't expect surprises since it's quite straightforward. But I may have missed something.

liewegas commented on src/crush/CrushWrapper.cc:
```diff
@@ -1311,6 +1315,13 @@ void CrushWrapper::decode(bufferlist::iterator& blp)
+      for (auto &c : class_name)
+	class_rname[c.second] = c.first;
+      ::decode(class_bucket, blp);
+    }
```
this decodes class_bucket but it isn't encoded above.. in another patch maybe?

liewegas replied:
actually it looks like the decode for class_bucket goes in teh later patch

liewegas commented on src/crush/CrushCompiler.cc:
```diff
@@ -52,6 +52,13 @@ static void print_item_class(ostream& out, int t, CrushWrapper &crush)
+{
+  const char *c = crush.get_class_name(t);
+  if (c)
+    out << " class " << c;
```
nit: maybe this shouldn't silently print nothing if passed bad input, since the callers check

dachary replied:
fixed with a comment showing the non fatal error
```
static void print_class(ostream& out, int t, CrushWrapper &crush)
{
  const char *c = crush.get_class_name(t);
  if (c)
    out << " class " << c;
  else
    out << " # unexpected class " << t;
}
```

liewegas commented on src/crush/CrushCompiler.cc:
```diff
@@ -270,7 +277,19 @@ int CrushCompiler::decompile(ostream &out)
+          int c;
+          int res = crush.split_id_class(step_item, &original_item, &c);
+          if (res < 0)
+            return res;
```
if res < 0, doesn't that mean we didn't split, and this is a normal bucket with no class specified?

liewegas replied:
ah, res < 0 is a real error, nevermind.

liewegas commented on src/crush/CrushCompiler.cc:
```diff
@@ -668,6 +671,8 @@ int CrushCompiler::parse_bucket(iter_t const& i)
 
 int CrushCompiler::parse_rule(iter_t const& i)
 {
+  crush.populate_classes();
```
maybe it's better to instead put this in parse_crush, guarded by a bool did_first_rule; or similar, so it's only done once before the first rule (and at the end if there were no rules).

dachary replied:
although uncommon the buckets may be defined in between rules
```
      // the whole crush map
      crushmap = *(tunable | device | bucket_type) >> *(bucket | crushrule);
```

liewegas commented on src/mon/OSDMonitor.cc:
```diff
@@ -4551,6 +4551,7 @@ int OSDMonitor::crush_rename_bucket(const string& srcname,
 
   pending_inc.crush.clear();
+  newcrush.cleanup_classes();
   newcrush.encode(pending_inc.crush, mon->get_quorum_con_features());
```
These seem very fragile because the map is unusable after this.  (Same with decompile_crush(), although there it's rebuilt so just wasted work).  Perhaps instead it's better to make encode smart enough to skip all the extra buckets?

dachary replied:
Hum. The
```
      cleanup_classes();
      populate_classes();
```
in CrushWrapper::decode should actually be
```
      cleanup_classes();
```
and we can get rid of mots the other calls in OSDMonitor. 

case a) A non device classes aware client gets a crushmap from a device class aware mon, modify it, encode it, send it back to a device class aware mon. OK

case b) A non device class aware client gets a crushmap from a device aware mon including generated buckets, and calls a non device class aware crushtool to decompile it. It will fail to recompile because of the special bucket names. A better error message suggesting an upgrade should be displayed in this case.

case C) A device class aware client sends a crushmap to a non device class aware mon (maybe because it is the leader of a cluster where an upgrade is in progress) the mon will be able to use the crushmap exactly as a device class aware mon would. Only it will never trim unused buckets with special names.

The ```crush.populate_classes();``` in CrushCompiler::decompile should be removed.
The ```  crush.populate_classes();``` in CrushCompiler::parse_crush should really be cleanup_classes since populate_classes() is called before parsing a rule.
All cleanup_classes from OSDMonitor should be removed since it is called by decode anyway.
The cleanup_classes from crushtool.cc should be removed since it is is called from CrushCompiler::parse_crush and from CrushCompiler::decompile

I'll do these changes if they makes sense to you. Or if it's all too confusing and implementing them will help clarify ;-)

liewegas commented:
Perhaps we can limit the callers to just finalize(), decode(), and 
compile() then?  Would that work?

dachary commented:
> Perhaps we can limit the callers to just finalize(), decode(), and compile() then? Would that work?
I'll try that, I think it's the way to go.

dachary commented:
jenkins test this please (ceph_objectstore_tool.py failed)

dachary commented:
I reworked the "crush: update device classes where relevant" commit and it's now much simpler. Also updated the comment of this commit to explain how it works and why it is compatible with existing clients in detail.

dachary commented:
<code>teuthology-suite -k distro --priority 1000 --suite rados --subset $(expr $RANDOM % 200)/200 --email loic@dachary.org --ceph wip-18943-device-class --machine-type smithi</code>

* **running** http://pulpito.ceph.com/loic-2017-03-01_15:46:12-rados-wip-18943-device-class-distro-basic-smithi

dachary commented:
<code>teuthology-suite -k distro --priority 1000 --suite rados --subset $(expr $RANDOM % 200)/200 --email loic@dachary.org --ceph wip-18943-device-class --machine-type smithi</code>

* **running** http://pulpito.ceph.com/loic-2017-03-02_10:34:12-rados-wip-18943-device-class-distro-basic-smithi

liewegas commented:
\o/
